internal/error.go

```diff
@@ -369,3 +479,260 @@ func newUnknownExternalWorkflowExecutionError() *UnknownExternalWorkflowExecutio
+	if errors.As(err, &terminatedErr) || errors.As(err, &canceledErr) || errors.As(err, &workflowPanicErr) {
+		return false
+	}
+
+	var applicationErr *ApplicationError
+	var applicationErrOriginalType string
+	if errors.As(err, &applicationErr) {
+		if applicationErr.nonRetryable {
+			return false
+		}
+		applicationErrOriginalType = applicationErr.orginalType
+	}
+
+	var timeoutErr *TimeoutError
+	if errors.As(err, &timeoutErr) {
+		if timeoutErr.timeoutType != commonpb.TimeoutType_StartToClose &&
+			timeoutErr.timeoutType != commonpb.TimeoutType_Heartbeat {
+			return false
+		}
+	}
+
+	var serverErr *ServerError
+	if errors.As(err, &serverErr) {
+		if serverErr.nonRetryable {
+			return false
+		}
+	}
+
+	for {
+		causeErr := errors.Unwrap(err)
+		if causeErr == nil {
+			break
+		}
+		err = causeErr
+	}
+	errType := getErrorType(err)
+	for _, nonRetryableType := range nonRetryableTypes {
+		if nonRetryableType == errType || nonRetryableType == applicationErrOriginalType {
+			return false
+		}
+	}
+
+	return true
+}
+
+func getErrorType(err error) string {
+	var t reflect.Type
+	for t = reflect.TypeOf(err); t.Kind() == reflect.Ptr; t = t.Elem() {
+	}
+
+	return t.Name()
+}
+
+// convertErrorToFailure converts error to failure.
+func convertErrorToFailure(err error, dc DataConverter) *failurepb.Failure {
+	if err == nil {
+		return nil
+	}
+
+	failure := &failurepb.Failure{
+		Source:  "GoSDK",
+		Message: err.Error(),
+	}
+
+	switch err := err.(type) {
+	case *ApplicationError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrorType(err),
+			NonRetryable: err.nonRetryable,
+			Details:      convertErrDetailsToPayloads(err.details, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+	case *CanceledError:
+		failureInfo := &failurepb.CanceledFailureInfo{
+			Details: convertErrDetailsToPayloads(err.details, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_CanceledFailureInfo{CanceledFailureInfo: failureInfo}
+	case *PanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type: getErrorType(err),
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *workflowPanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrorType(&PanicError{}),
+			NonRetryable: true,
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *TimeoutError:
+		failureInfo := &failurepb.TimeoutFailureInfo{
+			TimeoutType:          err.timeoutType,
+			LastFailure:          convertErrorToFailure(err.lastErr, dc),
+			LastHeartbeatDetails: convertErrDetailsToPayloads(err.lastHeartbeatDetails, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_TimeoutFailureInfo{TimeoutFailureInfo: failureInfo}
+	case *TerminatedError:
```

Comment:
In what situation we will get this failure from the application?  Looks like dead code to me.

Reply:
Yes, you are probably right. But as soon as we have both `TerminatedFailureInfo` and `TerminatedError` I decided to leave this 1:1 mapping. Can `TerminatedError` be the cause for, let's say, child workflow failure? Then it might be called as part of `cause` conversion.